File: war.py
from random import shuffle
import time, sys

SUITS = ["spades", "clubs", "hearts", "diamonds"]
hand1 = []
hand2 = []
dis1 = []
dis2 = []
curr = []
p = 0
p1 = 0
p2 = 0
wars = 0

# Creates the deck
def createDeck():
	deck = []
	for suit in SUITS: #Create the deck
		for i in range(1,14):
			deck.append([i,suit])
	shuffle(deck) #Shuffle the deck
	while(len(deck) is not 0): #Deal the cards to each player
		hand1.append(deck.pop())
		hand2.append(deck.pop())

# Plays through a game
def playGame():
	global p1, p2
	turn = 0
	
	while(checkDeck(0) == 0):
		if (turn > 3000): #Make sure the game does not loop infinitely
			break
		turn = turn + 1
		curr.append(hand1.pop()) #Add P1 card to play
		curr.append(hand2.pop()) #Add P2 card to play
		compareNpay()

	if(checkDeck(0) == 2):
		p1 += 1
	else:
		p2 += 1
	return turn

# Compare the cards and decide a winner 
def compareNpay():
		if (curr[0][0] == curr[1][0]):
			war()
		elif (curr[0][0] > curr[1][0]):
			while(len(curr) != 0):
				shuffle(dis1)
				dis1.append(curr.pop())
		else:
			while(len(curr) != 0):
				shuffle(dis2)
				dis2.append(curr.pop())
		checkDeck(0)

# Verify the deck has enough cards
def checkDeck(force):
	rt = 1 if (len(hand1) == 0 and len(dis1) == 0) else 2 if (len(hand2) == 0 and len(dis2) == 0) else 0

	if (len(hand1) == 0 or force %2 == 1):
		while(len(dis1) != 0):
			hand1.append(dis1.pop())

	if (len(hand2) == 0 or force >= 2):
		while(len(dis2) != 0 ):
			hand2.append(dis2.pop())
	return rt

# Called to settle a tie
def war():
	global wars
	wars += 1
	tst = 3 if (len(hand1) <= 3 and len(hand2) <= 3) else 1 if (len(hand1) <= 3) else 2 if (len(hand2) <= 3) else 0
	if(checkDeck(tst) == 0):
		tie = 3 if (min(len(hand1),len(hand2)) > 4) else min(len(hand1),len(hand2))-1
		for i in range(0,tie):
			curr.append(hand1.pop())
			curr.append(hand2.pop())
		curr.insert(0,hand1.pop())
		curr.insert(0,hand2.pop())
		compareNpay()

def main():
	global wars, p1, p2
	avg = 0
	start_time = time.time()
	try:
		if sys.argv[1].isdigit():
			times = int(sys.argv[1])
	except:
		times = 100
	for i in range(0,times):
		createDeck()
		avg += playGame()
		hand1.clear()
		hand2.clear()
		dis1.clear()
		dis2.clear()
	avg = avg/int(times)
	print(str(times) + ' games')
	print('Average turns: ' + str(int(avg)))
	print('P1: ' + str(p1) + '\tP2: ' + str(p2))
	print('Avg wars per game: ' + str(int(wars/times)))
	print('time to run: ' + str((time.time() - start_time)))

if __name__ == '__main__':
	main()

File: war.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>
#include <thread>

using namespace std;
using namespace std::chrono;

struct card {
	char suit;
	int value;
};

vector<card> createDeck(){
	char suits[4] = {'s','c','h','d'};
	vector<card> deck;
	for (int i=0; i<4; i++){
		for(int j=1; j<14; j++){
			card temp = {suits[i],j};
			deck.push_back(temp);
		}
	}
	return deck;
}

int winner(vector<card> play){
	if (play[0].value == play[1].value){return 2;}//Draw
	else if (play[0].value > play[1].value ){return 0;}//P1 Wins
	else {return 1;}//P2 Wins
}

int playGame(){
    vector<card> deck = createDeck();
	vector<card> p1;
	vector<card> p2;
	vector<card> war;
	int turns=0, bet;
	vector<card> play;
	//Random Shuffle cards
	srand(time(NULL));
	random_shuffle(deck.begin(), deck.end());
	//Deal to each player
	for (int i=0;i<deck.size();i+=2){
		p1.push_back(deck[i]);
		p2.push_back(deck[i+1]);
	}
	//Loop through til someone runs out of cards
	while(p1.size() != 0 && p2.size() != 0 && turns < 3000){
		//Put cards in play
		play.insert(play.begin(), 1, p1.front());
		play.insert(play.begin(), 1, p2.front());
		//Take cards off each players deck
		p1.erase(p1.begin());
		p2.erase(p2.begin());
        //cout << "P1: " << play[0].suit << play[0].value << "(" << p1.size() << ") P2: " << play[1].suit << play[1].value << " (" << p2.size() << ") Turns: " << turns << endl;
		switch (winner(play)){
			//P1 Wins
			case 0:
				p1.insert(p1.end(), play.begin(), play.end());
				play.clear();
                turns++;
			  break;
			//P2 Wins
			case 1: 
				p2.insert(p2.end(), play.begin(), play.end());
				play.clear();
                turns++;
		    break;

			//Draw, go to war
			case 2:
			  //Take the top 3 cards from each player as a bet for next round
				if (p1.size() < 3 || p2.size() < 3){
					bet = min(p1.size(), p2.size()) - 1;
				}
				else {bet = 3;}
			  for(int i=0; i<bet; i++){
					play.insert(play.begin(), 1, p1.front());
					play.insert(play.begin(), 1, p2.front());
					p1.erase(p1.begin());
					p2.erase(p2.begin());
			  }
              turns++;
		    break;
		}
	}
	if (p1.size() == 0){
		return 1;
	}
	else if (p2.size() == 0){
		return 0;
	}
    else{
        return 2;
    }
}

int main(int argc, char* argv[]){
    if (argc <=1){
        cout << "You need to tell me how many games!\n";
        exit(1);
    }
    
    int p1=0, p2=0, out, times=0, draw=0;
    auto start = high_resolution_clock::now(); 
    for (int i=0; i<atoi(argv[1]); i++){
        out = playGame();
        if (out == 1){p2++;}
        else if (out == 0){p1++;}
        else {draw++;}
    }
    auto stop = high_resolution_clock::now();
    auto duration = duration_cast<microseconds>(stop - start);
    cout << "(P1) Games Won: " << p1 << endl;
    cout << "(P2) Games Won: " << p2 << endl;
    cout << "Games exceeded 3000 turns: " << draw << endl;
    cout << "Time to run: " << duration.count() /1000000.00 << " seconds\n";
	return 0;
}